src/core/execute.c

```diff
@@ -3044,6 +3064,29 @@ static void invalid_env(const char *p, void *userdata) {
         log_unit_error(info->unit, "Ignoring invalid environment assignment '%s': %s", p, info->path);
 }
 
+int exec_context_named_iofds(Unit *unit, const ExecContext *c, const ExecParameters *p, int named_iofds[3]) {
+        unsigned i, targets;
+        const char *stdin_fdname, *stdout_fdname, *stderr_fdname;
+        stdin_fdname = (c->std_input_fdname ?: "stdin");
```

Comment:
i think it would make sense to add a new function that does the above check, and use that here and in `property_get_input_fdname()`, i.e. something like this added to execute.h:

```c
const char* exec_context_fdname(const ExecContext *c, int fd_index) {
        if (fd_index == STDIN_FILENO) {
                if (c->std_input != EXEC_INPUT_NAMED_FD)
                        return NULL;
                return c->std_input_fdname ?: "stdin";
        } else …
}
```

You get the idea...